### PR TITLE
Count API updates

### PR DESCRIPTION
**Pull request by rozza**
count() method deprecated
estimatedDocumentCount() methods added
documentCount() methods added

JAVA-2885

### CODE REVIEW

File: driver-core/src/main/com/mongodb/client/model/DocumentCountOptions.java [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+import java.util.concurrent.TimeUnit;
+
+import static com.mongodb.assertions.Assertions.notNull;
+
+/**
+ * The options for a counting documents in a collection.
+ *
+ * @since 3.8
+ */
+public class DocumentCountOptions {
```
Comment (rozza) [inline driver-core/src/main/com/mongodb/client/model/DocumentCountOptions.java]: This has been added so we don't surface a `hintString` option that can't be used in the aggregation command.

File: driver-async/src/main/com/mongodb/async/client/MongoCollection.java [hunk 2]
```diff
@@ -156,15 +158,19 @@
...
      */
+    @Deprecated
     void count(SingleResultCallback<Long> callback);
 
     /**
      * Counts the number of documents in the collection according to the given options.
      *
      * @param filter   the query filter
      * @param callback the callback passed the number of documents in the collection
+     * @deprecated use {@link #countDocuments(Bson, SingleResultCallback)} instead
```
Comment (jyemin) [inline driver-async/src/main/com/mongodb/async/client/MongoCollection.java]: I have a slight preference for the first doc comment just above, which mentions both of the new methods

File: driver-core/src/main/com/mongodb/operation/CountOperation.java [hunk 3]
```diff
@@ -297,4 +354,36 @@ private BsonDocument getCommand(final SessionContext sessionContext) {
         }
         return document;
     }
+
+    private AggregateOperation<BsonDocument> getAggregateOperation() {
+        if (hint != null && !hint.isDocument()) {
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/operation/CountOperation.java]: This is supportable by aggregate command:

MongoDB Enterprise repl0:PRIMARY> db.test.getIndexes()
[
	{
		"v" : 2,
		"key" : {
			"_id" : 1
		},
		"name" : "_id_",
		"ns" : "test.test"
	}
]
MongoDB Enterprise repl0:PRIMARY> db.runCommand({aggregate : "test", pipeline : [{$match : {_id : 1}}], cursor : {}, hint : "_id_"})
{
	"cursor" : {
		"firstBatch" : [ ],
		"id" : NumberLong(0),
		"ns" : "test.test"
	},
	"ok" : 1,
	"operationTime" : Timestamp(1529329089, 1),
	"$clusterTime" : {
		"clusterTime" : Timestamp(1529329089, 1),
		"signature" : {
			"hash" : BinData(0,"AAAAAAAAAAAAAAAAAAAAAAAAAAA="),
			"keyId" : NumberLong(0)
		}
	}
}

File: driver-core/src/main/com/mongodb/client/model/CountStrategy.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+ */
+
+package com.mongodb.client.model;
+
+/**
+ * The strategy to use for calculating the count value.
+ *
+ * @since 3.8
+ */
+public enum CountStrategy {
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-core/src/main/com/mongodb/client/model/CountStrategy.java]: Done

File: driver-core/src/main/com/mongodb/client/model/DocumentCountOptions.java [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+
+import com.mongodb.lang.Nullable;
+import org.bson.conversions.Bson;
+
+import java.util.concurrent.TimeUnit;
+
+import static com.mongodb.assertions.Assertions.notNull;
+
+/**
+ * The options for a counting documents in a collection.
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-core/src/main/com/mongodb/client/model/DocumentCountOptions.java]: Done

File: driver-core/src/main/com/mongodb/client/model/EstimatedDocumentCountOptions.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+ */
+
+package com.mongodb.client.model;
+
+import java.util.concurrent.TimeUnit;
+
+import static com.mongodb.assertions.Assertions.notNull;
+
+/**
+ * The options for a count operation.
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-core/src/main/com/mongodb/client/model/EstimatedDocumentCountOptions.java]: Done.

File: driver-core/src/main/com/mongodb/operation/CountOperation.java (hunk 3 above)
Comment (rozza) [reply driver-core/src/main/com/mongodb/operation/CountOperation.java]: Ack - updated.

**Comment by jyemin** [comment]
Travis build is failing, so probably a checkstyle failure snuck in the last commit.

**Pull request closed without merging**